web/src/main/scala/quasar/api/services/data.scala

```diff
@@ -0,0 +1,110 @@
+
+  implicit val naturalParamDecoder: org.http4s.QueryParamDecoder[Natural] = new QueryParamDecoder[Natural] {
+    def decode(value: QueryParameterValue): ValidationNel[ParseFailure, Natural] =
+      QueryParamDecoder[Long].decode(value).flatMap(long =>
+        Natural(long).toSuccess(NonEmptyList(ParseFailure(value.value, "must be >= 0")))
+      )
+  }
+
+  implicit val positiveParamDecoder: org.http4s.QueryParamDecoder[Positive] = new QueryParamDecoder[Positive] {
+    def decode(value: QueryParameterValue): ValidationNel[ParseFailure, Positive] =
+      QueryParamDecoder[Long].decode(value).flatMap(long =>
+        Positive(long).toSuccess(NonEmptyList(ParseFailure(value.value, "must be >= 1")))
+      )
+  }
+
+  object Offset extends OptionalValidatingQueryParamDecoderMatcher[Natural]("offset")
+  object Limit  extends OptionalValidatingQueryParamDecoderMatcher[Positive]("limit")
+
+  type FSTask[A] = FileSystemErrT[Task, A]
+
+  val unstack = new (FSTask ~> Task) {
+    def apply[A](t: FSTask[A]): Task[A] =
+      t.fold(e => Task.fail(new RuntimeException(e.shows)), Task.now).join
+  }
+
+  def service[S[_]: Functor](f: S ~> Task)(implicit R: ReadFile.Ops[S],
+                                                    W: WriteFile.Ops[S],
+                                                    M: ManageFile.Ops[S]): HttpService = {
+
+    def linesResponse[A: EntityEncoder](v: Process[R.M, A]): Task[Response] = {
+      val trans: R.F ~> Task = hoistFree(f)
+      val trans2: R.M ~> FSTask  = Hoist[FileSystemErrT].hoist(trans)
+      new PrOps(v.translate(trans2)).unconsOption.fold(
+        fileSystemErrorResponse,
+        _.fold(Ok(""))({ case (first, rest) =>
+          Ok(Process.emit(first) ++ rest.translate(unstack))
+        })).join
+    }
+    HttpService {
+      case req @ GET -> AsPathyPath(path) :? Offset(offsetParam) +& Limit(limitParam) => {
+        val offsetWithDefault = offsetParam.getOrElse(Natural._0.successNel)
+        val offsetWithErrorMsg: String \/ Natural = offsetWithDefault.disjunction.leftMap(
+          nel => s"invalid offset: ${nel.head.sanitized} (${nel.head.details})")
+        val limitWithErrorMsg: String \/ Option[Positive] = limitParam.traverseU(_.disjunction.leftMap(
+          nel => s"invalid limit: ${nel.head.sanitized} (${nel.head.details})"))
+        val possibleResponse = for {
+          offset <- offsetWithErrorMsg
+          limit <- limitWithErrorMsg
+        } yield {
+          val accept = req.headers.get(Accept)
+
+          path.fold(
+            dirPath => {
+              val format = MessageFormat.fromAccept(accept)
+              val zippedBytes = retrieveDirBytes[S](dirPath, format, offset, limit)
+              val contentType = `Content-Type`(MediaType.`application/zip`)
+              val headers = contentType :: format.disposition.toList
+              linesResponse(zippedBytes).map(_.replaceAllHeaders(headers: _*))
+            },
+            filePath => ???
```

Comment:
ya,ya :smile: Still in progress, wanted to get a PR up anyway. I actually got to this before lunch